### PR TITLE
Update th/td to use style attribute

### PR DESCRIPTION
**Pull request by gaige**
This allows better interoperation with CSS style sheets, as the align
object on the TH is skipped if the css uses 'text-align: inherit' and
the previous 'text-align' is used instead (or the default: left).

With the existing `align` attribute, standard CSS libraries, such as Bootstrap, can ignore the locally-attached `align` attribute by incorporating `text-align: inherit`. By setting the `text-align` CSS attribute in a `style` attribute attached to the th/td element, the local attribute overrides the CSS.  Without it, the lower-priority `align` attribute is ignored in these circumstances.

This is my first contribution to this project, so please let me know if there's something else you'd like. I think the commit messages, test updates, and change log updates are per spec.

### CODE REVIEW

**Comment by waylan** [comment]
Interesting. However, this extension is a strict clone of the PHP Markdown Extra implementation of tables. We expect our output to match that libraries’ output. This would change that. Are there any other implementations which use style attributes rather than align attributes for tables?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by gaige** [comment]
`align` and `valign` (along with attributes that directly manipulated display attributes) were deprecated in HTML4.  The current PHP implementation has 2 methods of doing alignment, the default is to still use the `align` attribute, but conditionally (with the definition of `table_align_class_tmpl`), it will insert a `class='$table_align_class_tmpl-{left|right|center}'`, which would allow for a somewhat more flexible implementation.  Unfortunately, that requires the introduction of an additional class, but if you'd prefer something like that, I could re-implement with that option. 

In that case, I'd be adding a config item to tables for `ALIGN_CLASS_TMPL` which would default to `None` and, if set, would add `class="<align_class>"` after replacing the `%%` in the string with the `left`, `right`, or `center` string.

That will stick most closely to the implementation in https://github.com/michelf/php-markdown.  I'm not entirely sure where to implement the test, although it appears test_syntax/extensions/test_tables might be a home for it.

Thoughts?

File: markdown/extensions/tables.py
Comment (mitya57) [inline markdown/extensions/tables.py]: The only thing I would suggest is using a space in `style` attribute, and maybe adding a trailing semicolon.
```suggestion
                c.set('style', f'text-align: {a};')
```

**Review by mitya57 — COMMENTED** [review]
I agree with @gaige — the attribute way of specifying alignment was deprecated in HTML 4 and the HTML 5 specification does not even mention it:
https://www.w3.org/TR/html53/tabular-data.html#attributes-common-to-td-and-th-elements

Mozilla’s documentation lists it with “should no longer be used” mark:
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/td#attr-align

I think there is no need to add a configuration option as the CSS `text-align` property should be supported by all major browsers:
https://caniuse.com/mdn-css_properties_text-align

**Comment by gaige** [comment]
@waylan Any thoughts about my comments and/or @mitya57 comments?  Anything I can do to help move this forward?

**Comment by waylan** [comment]
> current PHP implementation has 2 methods of doing alignment

That helps tremendously. Quite frankly I was surprised they were still using the `align` attribute. That said, technically, the original Markdown rules require the output to be XHTML. Of course, XHTML like syntax is a valid form of HTML5 so we mostly ignore the XHTML output requirement except that the tags take on that form (`<hr />` as opposed to `<hr>`). No doubt this is where the current default of using the `align` attribute comes from.

We may still have some users who are expecting the older formats and are using the related doctypes in their documents. Would changing this cause their pages to no longer be valid? True XHTML (not XHTML served as HTML) is especially problematic in this regard and can cause a browser to crash hard if it is invalid (i don't recall if `style` attributes are valid or not). If there is any chance that some users are doing so (and I suspect there is), then we would need to have a configuration option to allow users to select there preferred option if we make any changes at all.

Another consideration are (X)HTML sanitizers. For example, GitHub's sanitizer strips all `style` attributes out of all user supplied content on their site(s). Therefore, they are still [using][1] `align` attributes in their output. True, our output is not being run through GitHub's sanitizer, but it certainly could be run through similarly configured sanitizers by some of our users.

Given the above, I think a configuration option for the extension makes the most sense. I think we could probably change the default and place a big warning in the release notes informing users that if they want/need `align` attributes rather than `style` attributes, then they need to set a custom option on the extension.

> I'm not entirely sure where to implement the test, although it appears test_syntax/extensions/test_tables might be a home for it.

Yes, that is correct, In fact, the existing tests should be moved there as part of this change. Our policy has been to move any related tests to the new system when the code related to those tests is updated.

[1]: https://github.github.com/gfm/#example-199

File: markdown/extensions/tables.py
Comment (waylan) [reply markdown/extensions/tables.py]: Yes, I agree with all three changes here. Add the space and the semicolon and use an f-string, which is the most performant way to combine strings.

**Comment by gaige** [comment]
@waylan Changes have been made per discussion. Please let me know if there are any other changes.  I made a notation about the change, but perhaps there's either a better place to warn about it, or somewhere else you'd prefer, please let me know.

Also, checklinks is failing, but not not in files that I've been working in. Not sure what you want to do about that, but those are the only failing checks at this point.

File: docs/change_log/index.md
Comment (waylan) [inline docs/change_log/index.md]: As this is a backward incompatible change, this can't be included in a point release (see the [Contributing Guide][0]). It will need to wait until the next MINOR release (version 3.4.0) and the note will go in the Backward Incompatible Changes section of a separate document (see the release notes for [version 3.3][1] for an example. The note can then be more lengthy and perhaps include an example of how to set the config option to restore the old behavior.

[0]: https://python-markdown.github.io/contributing/#versions
[1]: https://github.com/Python-Markdown/markdown/blob/master/docs/change_log/release-3.3.md

File: tests/extensions/extra/tables.html
Comment (waylan) [inline tests/extensions/extra/tables.html]: Any tests which have been replicated in `tests/test_syntax/extensions/test_tables.py` should be removed from this file. If that leaves this file empty, then it can be deleted long with the entry for it in [tests/test_legacy.py](https://github.com/Python-Markdown/markdown/blob/master/tests/test_legacy.py#L185)

File: tests/test_syntax/extensions/test_tables.py
Comment (waylan) [inline tests/test_syntax/extensions/test_tables.py]: You can still wrap this in `self.dedent`, which is easier to read. The first line sets the amount of indentation to remove. As it is outdented from the table, it will work correctly.

Same goes for a number of other tests below.

**Review by waylan — CHANGES_REQUESTED** [review]
This is a great start. I have pointed out a few things below.

You can ignore the failing checklinks test. We are aware of the problem and need to resolve it separately.

File: docs/change_log/index.md
Comment (gaige) [reply docs/change_log/index.md]: Moved to release 3.4

File: tests/test_syntax/extensions/test_tables.py
Comment (gaige) [reply tests/test_syntax/extensions/test_tables.py]: I've indented/dedented where possible. One specific line was too long in a later test and because it already contains a large number of `\`s, I opted to leave it flush for clarity instead of wrapping it with a `\` at the end of the line prior.

**Review by waylan — APPROVED** [review]

File: tests/test_syntax/extensions/test_tables.py
Comment (waylan) [inline tests/test_syntax/extensions/test_tables.py]: This is only 107 characters long. However, we allow up to 128 character line length. There is plenty of room to indent this.

**Comment by waylan** [comment]
This looks good. I will merge this once we are ready to start development on version 3.4.

File: tests/test_syntax/extensions/test_tables.py
Comment (gaige) [reply tests/test_syntax/extensions/test_tables.py]: I had indented this line before I sent it in. However, running `tox` against it blew a warning on the line if I didn't split it, which is why I pulled it back flush. I'm not sure if that's a tox problem or something else, but I figured that if `tox` was complaining I should fix it. As mentioned previously, if there weren't already so many escapes in the line, I would have quoted the newline and left it indented, but that didn't seem a good idea here.

Comment (mitya57) [reply tests/test_syntax/extensions/test_tables.py]: > if there weren't already so many escapes in the line

Maybe you can use a raw string here (`r"""`) to reduce the number of escapes?

Comment (gaige) [reply tests/test_syntax/extensions/test_tables.py]: Good call. That resolved it. Code updated.

File: docs/change_log/release-3.4.md
Comment (crisluengo) [inline docs/change_log/release-3.4.md]: Typo: "TableExtensioon" -> "TableExtension"

Comment (gaige) [reply docs/change_log/release-3.4.md]: Thanks. Not sure how that happened, I thought I'd copied and pasted that from my other window.

**Pull request merged**
